### PR TITLE
[Android] Add device param of GetAttestationChallenge.

### LINKED ISSUE #17351: GetAttestationChallenge is a public API and doesn't make sense
#### Problem
GetAttestationChallenge ended up as a public API. BUT, it really is just a convenience for the commissioner to get the attestation challenge of the device CURRENTLY being commissioned because the proxy isn't provided. Previously, this (accidentally IMO) fell back to the most recently PASE'd proxy, but as we moved towards less statefulness, this function got left behind.

 #### Proposed Solution
Add optional var to the API to allow a proxy to be provided, THEN fall back to the current commissioning device.
CHIP_ERROR GetAttestationChallenge(ByteSpan & attestationChallenge, const DeviceProxy* proxy = nullptr)

### PR DESCRIPTION
**Pull request by chulspro**
Signed-off-by: sanghyukko <sanghyuk.ko@samsung.com>
Signed-off-by: Charles Kim <chulspro.kim@samsung.com>

#### Problem
We need to update info of device proxy during attestation step.

#### Change overview
Change to set device proxy value on attestation step and use device secure session.

#### Testing
Device commissioning

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -768,13 +768,13 @@ DeviceCommissioner::ContinueCommissioningAfterDeviceAttestationFailure(DevicePro
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR DeviceCommissioner::GetAttestationChallenge(ByteSpan & attestationChallenge)
+CHIP_ERROR DeviceCommissioner::GetAttestationChallenge(DeviceProxy * device, ByteSpan & attestationChallenge)
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:771]: Does this method now use `this` in any way? Should it be marked static in the header if it never uses this (and eventually moved around/refactored)?

**Review by andy31415 — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Review by msandstedt — CHANGES_REQUESTED** [review]
Instead of changing the `GetAttestationChallenge` interface, it would create less churn to add the desired `GetAttestationChallenge` signature as an overload.

The justification for this PR can only be that there are out-of-tree callers that need this, or that `mDeviceBeingCommissioned` is not the correct device for in-tree examples and we in fact have an underlying bug that we need to address directly.

Assuming the former (the need for this comes from out-of-tree callers), it does not makes sense to remove the existing interface, which can break other out-of-tree code.

An alternative would be just to manually extract the attestation challenge from the passed device pointer in the way this method demonstrates.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Yes, I don't understand this at all.  If we have a specific device, we should just be getting the attestation challenge from that (and adding API on DeviceProxy as needed for that purpose).  The only reason we have `GetAttestationChallenge` on the _controller_ is specifically to get the attestation challenge of `mDeviceBeingCommissioned`.

**Comment by chulspro** [comment]
> Instead of changing the `GetAttestationChallenge` interface, it would create less churn to add the desired `GetAttestationChallenge` signature as an overload.
> 
> The justification for this PR can only be that there are out-of-tree callers that need this, or that `mDeviceBeingCommissioned` is not the correct device for in-tree examples and we in fact have an underlying bug that we need to address directly.
> 
> Assuming the former (the need for this comes from out-of-tree callers), it does not makes sense to remove the existing interface, which can break other out-of-tree code.
> 
> An alternative would be just to manually extract the attestation challenge from the passed device pointer in the way this method demonstrates.

Okay, I will add new API for this and commit one more time after launch. Thank you for your feedback

**Comment by chulspro** [comment]
> Yes, I don't understand this at all. If we have a specific device, we should just be getting the attestation challenge from that (and adding API on DeviceProxy as needed for that purpose). The only reason we have `GetAttestationChallenge` on the _controller_ is specifically to get the attestation challenge of `mDeviceBeingCommissioned`.

Okay, I will add new API recommended by @msandstedt

File: src/controller/CHIPDeviceController.cpp [hunk 2]
```diff
@@ -781,6 +781,19 @@ CHIP_ERROR DeviceCommissioner::GetAttestationChallenge(ByteSpan & attestationCha
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR DeviceCommissioner::GetAttestationChallenge(DeviceProxy * device, ByteSpan & attestationChallenge)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:784]: Again, why is this an API on DeviceCommissioner?  It does not use the DeviceCommissioner in any way.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Pull request closed without merging**